server/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -47,6 +47,15 @@ public static <T> Iterator<T> concat(Iterator<? extends T>... iterators) {
         return new ConcatenatedIterator<T>(iterators);
     }
 
+    public static int size(Iterator<?> iterator) {
+        int count = 0;
+        while (iterator.hasNext()) {
+            iterator.next();
+            count++;
+        }
+        return count;
+    }
```

Comment:
Its use is somewhat limited for production use cases as this consumes the iterator.
I wonder if we should have a custom hamcrest matches that asserts the element count in iterator instead?